File: app/src/main/java/de/cineaste/android/util/MultiList.java
package de.cineaste.android.util;

import android.support.annotation.NonNull;
import android.util.LongSparseArray;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import de.cineaste.android.entity.MovieDto;

public class MultiList {

    private final LongSparseArray<MultiListEntry> movies;

    public MultiList() {
        this.movies = new LongSparseArray<>();
    }

    public void add(MovieDto movieDto) {
        long movieId = movieDto.getId();
        MultiListEntry entry = movies.get(movieId);

        if (entry == null) {
            movies.put(movieId, new MultiListEntry(movieDto));
        } else {
            movies.put(movieId, entry.increment());
        }
    }

    public void addAll(List<MovieDto> movieDtos) {
        for (MovieDto movieDto : movieDtos) {
            add(movieDto);
        }
    }

    public List<MultiListEntry> getMovieList() {
        List<MultiListEntry> entries = new ArrayList<>();
        for (int i = 0; i < movies.size(); i++) {
            entries.add(movies.valueAt(i));
        }

        Collections.sort(entries);

        return entries;
    }

    public class MultiListEntry implements Comparable<MultiListEntry> {
        private final MovieDto movieDto;
        private int counter;

        private MultiListEntry(MovieDto movieDto) {
            this.movieDto = movieDto;
            this.counter = 1;
        }

        private MultiListEntry increment() {
            this.counter++;
            return this;
        }

        public MovieDto getMovieDto() {
            return movieDto;
        }

        public int getCounter() {
            return counter;
        }

        @Override
        public int compareTo(@NonNull MultiListEntry multiListEntry) {
            return multiListEntry.counter - this.counter;
        }
    }
}

File: app/src/main/kotlin/de/cineaste/android/util/MultiList.kt
package de.cineaste.android.util

import android.util.LongSparseArray

import java.util.ArrayList

import de.cineaste.android.entity.movie.MovieDto

class MultiList {

    constructor()

    constructor(movieDtos: List<MovieDto>) {
        addAll(movieDtos)
    }

    private val movies: LongSparseArray<MultiListEntry> = LongSparseArray()

    fun clear() = movies.clear()

    fun size(): Int = movies.size()

    fun sortedList(): List<MultiListEntry> {
        val entries = ArrayList<MultiListEntry>()
        for (i in 0 until movies.size()) {
            entries.add(movies.valueAt(i))
        }

        val comparator: Comparator<MultiListEntry> = createComparatorForCounterAndTitle()

        entries.sortWith(comparator)

        return entries
    }

    private fun createComparatorForCounterAndTitle(): Comparator<MultiListEntry> {
        val byCountComparator: Comparator<MultiListEntry> = compareByDescending { it.counter }

        return byCountComparator.thenBy { entry -> entry.movieDto.title }
    }

    fun add(movieDto: MovieDto) {
        val movieId = movieDto.id
        val entry = movies.get(movieId)

        if (entry == null) {
            movies.put(movieId, MultiListEntry(movieDto))
        } else {
            movies.put(movieId, entry.increment())
        }
    }

    fun addAll(movieDtos: List<MovieDto>) {
        for (movieDto in movieDtos) {
            add(movieDto)
        }
    }

    data class MultiListEntry(
        val movieDto: MovieDto,
        var counter: Int = 1
    ) {
        fun increment(): MultiListEntry {
            this.counter++
            return this
        }
    }
}